### PR TITLE
Improve breadcrumb

### PR DESCRIPTION
**Pull request by ericdallo**
Improves on our current breadcrumb following some ideas from https://github.com/emacs-lsp/lsp-mode/pull/1899:

* Add project path on breadcrumb (like vscode), currently without any mouse support/action 
* Add mouse support for document symbols
  * mouse-1: go to symbol 
  * mouse-2: narrow to region (following @danielmartin [suggestion](https://github.com/emacs-lsp/lsp-mode/issues/1823#issuecomment-651365097))
* Add function to go to symbol and function to narrow via prefix args: `C-3 M-x lsp-breadcrumb-go-to-symbol`

image

### CODE REVIEW

File: lsp-mode.el [hunk 1]
```diff
@@ -2087,24 +2087,73 @@ The `:global' workspace is global one.")
...
-                              (concat symbol2-icon symbol2-name)
-                            symbol2-name))))
+                          (lsp--headerline-with-action symbol-to-append full-symbol-2))))
               symbols-hierarchy ""))
 
+(defun lsp--headerline-dirs-until-root (root-path path)
+  "Find recursively the folders until the project ROOT-PATH.
+PATH is the current folder to be checked."
+  (let ((cur-path (list (f-base path))))
+    (if (f-same? root-path (f-parent path))
```
Comment (kiennq) [inline lsp-mode.el:2139]: `lsp-f-same?`, unless you want to follow the file symlink

Comment (kiennq) [reply lsp-mode.el:2139]: Btw, will this function to be called frequently? Following the symlink is quite slow. So you may want to rewrite `f-parent` so it's not using `f-same?` too.

Comment (ericdallo) [reply lsp-mode.el:2139]: Good catch, implemented `lsp-f-parent`

File: lsp-mode.el [hunk 2]
```diff
@@ -2087,24 +2087,73 @@ The `:global' workspace is global one.")
...
 
-(defun lsp--headerline-build-string (symbols-hierarchy)
-  "Build the header-line from SYMBOLS-HIERARCHY."
+(lsp-defun lsp--headerline-breadcrumb-go-to-symbol ((&DocumentSymbol :selection-range (&Range :start)))
+  "Go to breadcrumb symbol."
+  (->> start
+       lsp--position-to-point
+       goto-char))
+
+(lsp-defun lsp--headerline-breadcrumb-narrow-to-symbol ((&DocumentSymbol :range (&Range :start :end)))
```
[the remark this answers is not in the document]
Comment (ericdallo) [reply lsp-mode.el:2106]: I didn't find any `&RangeToRegion`, it is on master?

[the remark this answers is not in the document]
Comment (yyoncho) [reply lsp-mode.el:2106]: Sorry, it is &RangeToPoint

**Comment by yyoncho** [comment]
Seems like the file name is missing on the screenshot? Is that the case? Can you add the filename icons as well? You may use helm-icons package as inspiration.

**Comment by ericdallo** [comment]
@yyoncho it is displaying the file name but without the extension, I'll fix it

**Comment by ericdallo** [comment]
I don't know for sure how to get the icon from the filename, I'll take a look into helm-icons

File: lsp-mode.el [hunk 3]
```diff
@@ -2087,24 +2097,72 @@ The `:global' workspace is global one.")
...
+      (append (lsp--headerline-dirs-until-root root-path (lsp-f-parent path)) cur-path))))
+
+(defun lsp--headerline-breadcrumb-build-prefix-string ()
+  "Build the prefix for breadcrumb."
+  (seq-reduce (lambda (last-dirs next-dir)
+                (format "%s %s %s"
+                        last-dirs
+                        (lsp--headerline-breadcrumb-arrow-icon)
+                        (propertize next-dir 'font-lock-face lsp-headerline-breadcrumb-face)))
+              (lsp--headerline-dirs-until-root (lsp--suggest-project-root) (buffer-file-name)) ""))
```
Comment (yyoncho) [inline lsp-mode.el:2159]: it should be lsp-workspace-root.

**Comment by yyoncho** [comment]
Looks good! IMO we might also introduce an option to narrow deeply nested paths(e. g. use only first letters of the directories) when the path is longer.

I am not 100% how jump should work - e. g. do we need to add a command to jump into directories, e. g. M-2 M-x lsp-breadcrump-jump-dir to jump in the root or we should have one command for jumping and threat files + symbols as one three.

**Comment by ericdallo** [comment]
I see... it looks to be a good idea to implement on next, we can keep discussing on the linked issue

**Comment by ericdallo** [comment]
I'll merge this and implement the icons on a next PR too.

**Comment by yyoncho** [comment]
And one more thing - we should have a flag to turn off the path and keep only symbols.

**Pull request merged**